File: Algo Problems/Backtracking/Playlist Problems/Chess Type/nQueen.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution{
    private: 
        vector<vector<int>> board;
        int size;
    public: 
        Solution(int n){
            board.resize(n, vector<int>(n, 0));
            size = n;
        }

        void displayBoard(){
            cout<<"Board ===================\n"<<endl;
            for(auto it: board){
                for(auto x: it) {
                    if(x==1) cout<<"Q"<<" ";
                    else cout<<"_"<<" ";
                }
                cout<<endl;
            }
            cout<<"===================\n"<<endl;
        }

        bool isSafe(int row, int col){
            for(int i=0;i<size;i++){
                if(board[row][i]==1) return false;
                if(board[i][col]==1) return false;
            }
            int i=row, j=col;
            while(i>=0 && j>=0){
                if(board[i][j]==1) return false;
                i--, j--;
            }
            i=row, j=col;
            while(i>=0 && j<size){
                if(board[i][j]==1) return false;
                i--, j++;
            }
            return true;
        }

        void util(int row){
            if(row == size){
                displayBoard();
                return;
            }
            for(int j=0;j<size;j++){
                if(isSafe(row, j)){
                    board[row][j] = 1;
                    util(row+1);
                    board[row][j] = 0;
                }
            }
        }

        void solve(){
            util(0);
        }
};

int main(){
    Solution s(6);
    cout<<"Hello World"<<endl;
    s.solve();
}

File: Algo_Problems/Backtracking/ChessType/NQueen.java
package Algo_Problems.Backtracking.ChessType;

public class NQueen {
    public static void main(String[] args) {
        int board[][] = new int[4][4];
        solve(board, 0);
    }

    static void solve(int board[][], int row) {
        int n = board.length;

        if (row == n) {
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    if (board[i][j] == 1) {
                        System.out.print(" Q ");
                    } else {
                        System.out.print(" _ ");
                    }
                }
                System.out.println("");
            }
            System.out.println("");
            return;
        }

        for (int col = 0; col < n; col++) {
            if (isSafe(row, col, board)) {
                board[row][col] = 1;
                solve(board, row + 1);
                board[row][col] = 0;
            }
        }
    }

    static boolean isSafe(int x, int y, int board[][]) {
        int n = board.length;
        for (int i = 0; i < n; i++) {
            if (board[x][i] == 1)
                return false;
            if (board[i][y] == 1)
                return false;
        }

        int i = x, j = y;

        while (i >= 0 && j < n) {
            if (board[i][j] == 1) {
                return false;
            }
            i--;
            j++;
        }
        i = x;
        j = y;
        while (i >= 0 && j >= 0) {
            if (board[i][j] == 1) {
                return false;
            }
            i--;
            j--;
        }

        return true;
    }
}
